internal/resolution/variablesources/installed_package.go

```diff
@@ -1,76 +1,74 @@
-	if len(resultSet) == 0 {
-		return nil, r.notFoundError()
-	}
-
-	// TODO: fast follow - we should check whether we are already supporting the channel attribute in the operator spec.
-	//       if so, we should take the value from spec of the operator CR in the owner ref of the bundle deployment.
-	//       If that channel is set, we need to update the filter above to filter by channel as well.
-	sort.SliceStable(resultSet, func(i, j int) bool {
-		return catalogsort.ByVersion(resultSet[i], resultSet[j])
-	})
-	installedBundle := resultSet[0]
-
-	upgradeEdges, err := r.successors(allBundles, installedBundle)
-	if err != nil {
-		return nil, err
-	}
-
-	// you can always upgrade to yourself, i.e. not upgrade
-	upgradeEdges = append(upgradeEdges, installedBundle)
-	return []deppy.Variable{
-		variables.NewInstalledPackageVariable(installedBundle.Package, upgradeEdges),
-	}, nil
-}
-
-func (r *InstalledPackageVariableSource) notFoundError() error {
-	return fmt.Errorf("bundleImage %q not found", r.bundleImage)
-}
-
-func NewInstalledPackageVariableSource(catalogClient BundleProvider, bundleImage string) (*InstalledPackageVariableSource, error) {
-	successors := legacySemanticsSuccessors
+// InstalledPackageVariables returns variables representing packages
+// already installed in the system.
+// Meaning that each BundleDeployment managed by operator-controller
+// has own variable.
+func InstalledPackageVariables(
+	allBundles []*catalogmetadata.Bundle,
+	bundleDeployments []rukpakv1alpha1.BundleDeployment,
+) ([]*olmvariables.InstalledPackageVariable, error) {
+	var successors successorsFunc = legacySemanticsSuccessors
 	if features.OperatorControllerFeatureGate.Enabled(features.ForceSemverUpgradeConstraints) {
 		successors = semverSuccessors
 	}
 
-	return &InstalledPackageVariableSource{
-		catalogClient: catalogClient,
-		bundleImage:   bundleImage,
-		successors:    successors,
-	}, nil
+	result := make([]*olmvariables.InstalledPackageVariable, 0, len(bundleDeployments))
+	processed := sets.Set[string]{}
+	for _, bundleDeployment := range bundleDeployments {
+		sourceImage := bundleDeployment.Spec.Template.Spec.Source.Image
+		if sourceImage == nil || sourceImage.Ref == "" {
+			continue
+		}
+
+		if processed.Has(sourceImage.Ref) {
+			continue
+		}
+		processed.Insert(sourceImage.Ref)
+
+		bundleImage := sourceImage.Ref
+
+		// find corresponding bundle for the installed content
+		resultSet := catalogfilter.Filter(allBundles, catalogfilter.WithBundleImage(bundleImage))
+		if len(resultSet) == 0 {
+			return nil, fmt.Errorf("bundleImage %q not found", bundleImage)
+		}
+
+		// TODO: fast follow - we should check whether we are already supporting the channel attribute in the operator spec.
+		//       if so, we should take the value from spec of the operator CR in the owner ref of the bundle deployment.
+		//       If that channel is set, we need to update the filter above to filter by channel as well.
```

Comment:
This is an old todo. I'm trying to move code around without significant refactoring.

But we have an issue related to this: https://github.com/operator-framework/operator-controller/issues/466

Reply:
Maybe I'm misunderstanding this TODO, but it also just seems like something we don't want to do?

The only purpose of the `Operator.spec.channel` is to restrict what bundles are possible candidates for the next resolution. Once a bundle is installed, it makes no difference which channel it "came" from. There is no concept like "this bundle deployment represents a bundle from a specific channel".

Reply:
> Maybe I'm misunderstanding this TODO, but it also just seems like something we don't want to do?

I just re-read it and now I think I misunderstood it: looks like it is not related to #466. But I still think that #466 is an issue that needs to be addressed (I replied ro you rcomment [here](https://github.com/operator-framework/operator-controller/issues/466#issuecomment-1785166959)).

I interpreted it like "we should do a beter job at finding the correct bundle for installed package and we could use channel for that... Maybe?".



> The only purpose of the `Operator.spec.channel` is to restrict what bundles are possible candidates for the next resolution. Once a bundle is installed, it makes no difference which channel it "came" from. There is no concept like "this bundle deployment represents a bundle from a specific channel".

Yes, that was my understanding too.

I think we should delete this comment to avoid further confusion, but keep #466 open.

@joelanford What do you think?